### ISSUE TITLE
trial needs something to check DeprecationWarning

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2626]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2626|
|Type|enhancement|
|Created|2007-05-04 15:47:28Z|

There is a method in test_process.ProcessTestCase that asserts that a function call produce a warning. It's a bit like util.supress, but with checking the content of the warning, and I think that would be great for the deprecation policy.

<details><summary>Searchable metadata</summary>

```
trac-id__2626 2626
type__enhancement enhancement
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1178293648000000 1178293648000000
changetime__1181226259000000 1181226259000000
version__None None
owner__ 
cc__exarkun cc__jml cc__therve cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

I've added something like this in the `setUpClass-2303-2` branch (currently stalled due to lack of gtk1 support in feisty).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Where in that branch is the relevant code?  I looked around for a minute but I didn't see any likely candidates.  I was going to split it out into a separate branch so we could go ahead with this without waiting for [#2303](https://github.com/twisted/twisted/issues/2303).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

`t.trial.util.collectWarnings`. It isn't quite what's described, but it's the first step.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Note that it's in the ireporter-2164 branch, no setUpClass-2303.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Ready to review in assert-warn-2626.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Looks nice.  A few things:

  * No coverage for the code which suppresses the !__warningregistry!__ for the function being tested (covering that probably just involves using ```assertWarn``` twice with the same function)
  * I think ```args``` and ```kwargs``` should be ```*args``` and ```**kwargs```.  If this is a practical problem for someone, we can fix it later (by defining all the positional arguments with *args).
  * ```assertWarns``` closes matches the naming style of ```assertRaises```
  * The function being tested might result in multiple warnings; there's no way to test such a function the way ```assertWarn``` works now.  It'd be fine to defer that functionality to another ticket/time though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

I've addressed the 3 points. The last functionality could be useful, but I don't think that's a high priority.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm set owner to @therve|
|-|-|

Reviewed the last change set. Looking good. Please merge.

Maybe create a ticket for the latter point raised by exarkun?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [20485]) Merge assert-warn-2626

Author: therve
Reviewer: ralphm, exarkun
Fixes [#2626](https://github.com/twisted/twisted/issues/2626)

Add an assertWarns method on unittest.TestCase that allows to check
a warning is emitted when a function is called.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
